Pass password through to puppet directly rather than sed /etc/shadow

cryptk opened the pull request:
Using sed on the /etc/shadow file limits what distributions this module can successfully set a password on (for example,  on my Ubuntu 16.04 test system, it could not).  This PR adds a fact which forms a list of all linux users with a UID over 1000 (non-system users).  If the user does not exist already, the user resource sends a password attribute, and if the user does already exist, then it does not submit a password attribute.

This makes the module much more distribution agnostic while preserving the ability for users to manage their own passwords.

cryptk commented:
As an added bonus, this will fix #46

cryptk commented:
Aside from the coverage tests not working in travis, it looks like this PR breaks a lot of the other tests (because it sets the password to '*' by default)... I am exploring options on how to fix this...

rnelson0 commented:
@cryptk Neat! This appears to be a much better way to handle this. I will fix the base tests this weekend and let you know when it's ready for a rebase.

Until then my only code question is, why exclude users below uid 100? It will soon be possible to manage System users (#54) and it appears that might negatively affect *new* system users who would be left with the default nil/disabled password.

bbriggs commented:
+1 to @rnelson0's observation, but I see now you've just pushed a commit for that.

cryptk commented:
@rnelson0 good point, I removed the restriction on UID's.  Also, I already fixed the spec tests, just needed to add a fact for linux_users so that the if block wouldn't fail to create the user.

Also, I commented on #54 with what I think is a solution to the coverage task error in travis.

cryptk commented:
@rnelson0 one thing to mention is that behavior does change with this PR.  Assuming that users have passwords set up with escaped /, then new boxes that are puppetted up will not get the same password set in /etc/shadow...

Would you like me to mention this somewhere in the README?

cryptk commented:
Also, the reason I set an explicit default on the password attribute is to prevent it from creating users with an empty password (unless the user explicitly sets it to an empty string).  Empty passwords are free logins for everybody :fearful:

cryptk commented on manifests/init.pp:
```diff
@@ -29,7 +29,7 @@
   $state,
   $groups,
-  $password,
+  $password            = '*',
```
I'm not sure if one is better than the other... I could go either way, do you have a preference?

cryptk commented on spec/defines/init_spec.rb:
```diff
@@ -22,9 +23,9 @@
       :home             => '/home/rnelson0',
       :groups           => ['group1', 'group2'],
       :password_max_age => 90,
+      :password         => '*',
```
Just to validate that the default param is indeed being passed through properly

cryptk commented on spec/defines/init_spec.rb:
```diff
@@ -5,13 +5,14 @@
       :state    => 'present',
       :groups   => ['group1', 'group2'],
-      :password => 'encryptedstring',
+#      :password => 'encryptedstring',
```
can do

cryptk replied:
I actually reverted the password parameter stuff back to original instead

rnelson0 commented:
@cryptk looks great! Thanks for your patience on this issue.
